[Split Build] Use single command to build both wheels

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126930
* __->__ #126590
* #126929
* #126813
* #126699
* #126826

Adds a `SPLIT_BUILD` option to build both the libtorch wheel and pytorch wheels at the same time. This option was added as a convenience and right now only supports "clean" and "install"

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update on "[Split Build] Use single command to build both wheels"


Adds a `SPLIT_BUILD` option to build both the libtorch wheel and pytorch wheels at the same time. This option was added as a convenience and right now only supports "clean" and "install"


[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

atalman commented on setup.py:
```diff
@@ -230,20 +235,8 @@ def _get_package_path(package_name):
     return None
 
 
-BUILD_LIBTORCH_WHL = os.getenv("BUILD_LIBTORCH_WHL", "0") == "1"
```
Can we unify this one and https://github.com/pytorch/pytorch/pull/126328 ? Since these lines seems to be added by #126328 ?

PaliC replied:
These two PRs are wholly unrelated. The only reason they're together in a stack is to not have a merge conflict. I can just remove this PR and CI should still pass if you want

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

huydhn commented on setup.py:
```diff
@@ -1124,11 +1117,58 @@ def print_box(msg):
+            or setup_cmd == "sdist"
+            or setup_cmd == "develop"
+        ):
+            raise RuntimeError(
```
I'm a bit confusing here, if this rase a RunTimeError, then the rest of the if statement would be skipped, wouldn't it?

atalman replied:
I guess this is unsupported use case, looking at the Error text.

huydhn commented on setup.py:
```diff
@@ -1124,11 +1117,58 @@ def print_box(msg):
+    global BUILD_PYTHON_ONLY
+    global PACKAGE_NAME
+
+    BUILD_LIBTORCH_WHL = os.getenv("BUILD_LIBTORCH_WHL", "0") == "1"
```
Nit: Is there a reason to bring the initialization `BUILD_LIBTORCH_WHL = os.getenv("BUILD_LIBTORCH_WHL", "0") == "1"` and `BUILD_PYTHON_ONLY = os.getenv("BUILD_PYTHON_ONLY", "0") == "1"` here instead of where they are defined above?  It's better to defined and init a variable in the same place if possible IMO

huydhn commented on setup.py:
```diff
@@ -1124,11 +1117,58 @@ def print_box(msg):
+    _main()
+
+
+def _main():
```
I would love to call this function something else if possible `_main` doesn't feel like a good name, may be `_build`?

atalman replied:
+1 for this, using ``_build`` sounds like a good idea

atalman commented on setup.py:
```diff
@@ -1124,11 +1117,58 @@ def print_box(msg):
+            BUILD_PYTHON_ONLY = True
+            sys.argv[1] = "clean"
+            PACKAGE_NAME = final_package_name
+            _main()
```
Will this work same way as ?:
```
BUILD_LIBTORCH_WHL=1 python setup.py install
python setup.py clean
BUILD_PYTHON_ONLY=1 python setup.py install
```

The code above is calling _main() 3 times. While code here is only 2 times.

atalman commented on setup.py:
```diff
@@ -1124,11 +1117,58 @@ def print_box(msg):
+            )
+
+            final_package_name = PACKAGE_NAME
+            PACKAGE_NAME = "libtorch"
```
What are the final 2 package names ? Where we will store them in ``https://download.pytorch.org/whl/nightly/`` - we want to make sure there do not come to conflict with current torch packages

atalman reviewed (changes requested):
Please see comments

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update on "[Split Build] Use single command to build both wheels"


Adds a `SPLIT_BUILD` option to build both the libtorch wheel and pytorch wheels at the same time. This option was added as a convenience and right now only supports "clean" and "install"


[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]
